File: app/repositories/candle/candle_repository.py
from sqlalchemy import select

from app.models.candle import Candle


class CandleRepository:

    def __init__(
        self,
        db
    ):

        self.db = db

    async def create(
        self,
        candle: Candle
    ):

        self.db.add(candle)

        await self.db.commit()

        await self.db.refresh(candle)

        return candle

    async def get_candles(
        self,
        symbol: str,
        timeframe: str,
        limit: int = 100
    ):

        query = (
            select(Candle)
            .where(
                Candle.symbol == symbol,
                Candle.timeframe == timeframe
            )
            .limit(limit)
        )

        result = await self.db.execute(
            query
        )

        return result.scalars().all()

File: backend/src/repositories/candleRepository.ts
import { prisma } from "../core/database";
import { Candle } from "@prisma/client";

export class CandleRepository {
  static async create(data: {
    id: string;
    symbol: string;
    timeframe: string;
    timestamp: Date;
    open: number;
    high: number;
    low: number;
    close: number;
    volume: number;
  }): Promise<Candle> {
    return prisma.candle.create({
      data,
    });
  }

  static async getCandles(symbol: string, timeframe: string, limit: number = 100): Promise<Candle[]> {
    return prisma.candle.findMany({
      where: {
        symbol,
        timeframe,
      },
      orderBy: {
        timestamp: "desc",
      },
      take: limit,
    });
  }
}
